### PR TITLE
Bug Fix: Defensively copy context entities

### PR DESCRIPTION
**Pull request by tyler-dodge**
# Description

Before this change, concurrent async tasks would all share the same instance of the entities list. This change makes it so they each get their own copy of the list.

This matters because the recorder modifies the `entities` list in place, which makes it so concurrent subtasks end up looking at the wrong item in the `entites` list when deciding the parent subsegment:

https://github.com/aws/aws-xray-sdk-python/blob/0f13101e4dba7b5c735371cb922f727b1d9f46d8/aws_xray_sdk/core/context.py#L90-L101

You can see this in the unit test added. Before this change the unit test would print out the following:

```
Subseg 1 ID: 5b1c1cf70db2e808
Subseg 2 ID: 4a138f09cd3c6813
Subseg 3 ID: f97c02073584be8a
Subseg 4 ID: efac5645672e1cd6
Subseg 5 ID: 6874f083c496d388
Subseg 6 ID: f447815d60c734d6
Subseg 7 ID: edf7d60ea3d6875e
Subseg 8 ID: d2a1980e83a569f4
Subseg 9 ID: 6a0f1059ca4a91b6
Subseg 10 ID: 75421c2a04106214

Parent Segment ID: a9037c7ccc92e0c8 # Correct!
Subseg parent ID: a9037c7ccc92e0c8 # Correct!
Subseg parent ID: 4a138f09cd3c6813 # WRONG (all the ones below are wrong)
Subseg parent ID: f97c02073584be8a
Subseg parent ID: efac5645672e1cd6
Subseg parent ID: 6874f083c496d388
Subseg parent ID: f447815d60c734d6
Subseg parent ID: edf7d60ea3d6875e
Subseg parent ID: d2a1980e83a569f4
Subseg parent ID: 6a0f1059ca4a91b6
Subseg parent ID: 75421c2a04106214
```

With this change, all the subsegments have the correct parent ID.

(enowell) You can also visualize this in X-Ray from my tests:

Before:

image

Subsegments are children of each other even though they are concurrent:

After:

image

Subsegments have the correct parent ID.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

~I think this possibly fixes the issue here:~

Fixes: #310
Fixes: #164

### CODE REVIEW

**Comment by NathanielRN** [comment]
Thanks @tyler-dodge! Just curious, do you have a screenshot of this failing or the bug you observed? I think this change makes sense but I think it would be useful to see it breaking. (Or maybe unit tests that show it works if that’s easier to add?).

**Comment by tyler-dodge** [comment]
I added a unit test for it in the latest force push that I also verified failed prior to this change.

File: tests/test_async_recorder.py [hunk 1]
```diff
@@ -43,6 +44,28 @@ async def test_capture(loop):
...
+                # Ensure that the task subsegments overlap
+                if counter < total_tasks:
+                    await event.wait()
+                else:
+                    event.set()
+                return subsegment.parent_id
+        tasks = [assert_task() for task in range(total_tasks)]
+        results = await asyncio.gather(*tasks)
+        for result in results:
+            assert result == segment.id
```
Comment (NathanielRN) [inline tests/test_async_recorder.py:67]: Just a suggestion on naming:

```suggestion
        subsegs_parent_ids = await asyncio.gather(*tasks)
        for subseg_parent_id in subsegs_parent_ids:
            assert subseg_parent_id == segment.id
```

**Review by NathanielRN — APPROVED** [review]
Thanks so much for your help Tyler! This is a great bug fix and I look forward to seeing it being deployed 🙂 

I added some suggestions for documentation to help us in the future, would you mind taking a look? Thanks!

**Comment by tyler-dodge** [comment]
The modifications you added all look great to me.

**Pull request merged**